File: bindgen-tests/swift/tests/interfaces.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import uniffi_bindgen_tests

let interface = TestInterface.init(value: 20);
assert(interface.getValue() == 20);
assert(cloneInterface(interface: interface).getValue() == 20);

// Test records that store interfaces
//
// The goal is to test if we can read/write interface handles to RustBuffers
let two = TwoTestInterfaces(
  first: TestInterface.init(value: 1),
  second: TestInterface.init(value: 2)
);
let swapped = swapTestInterfaces(interfaces: two);
assert(swapped.first.getValue() == 2);
assert(swapped.second.getValue() == 1);

// Create 2 references to an interface using a bunch of intermediary objects:
//   * The one passed to `funcThatClonesInterface`
//   * The clones created for each method call

let interface2 = TestInterface.init(value: 20);
func funcThatClonesInterface(interface: TestInterface) -> TestInterface {
  return cloneInterface(interface: interface);
}
let interface2Clone = funcThatClonesInterface(
    interface: cloneInterface(interface: interface)
);
let _ = interface.getValue();
// Check that only the 2 actual references remain after the dust clears
assert(interface.refCount() == 2);

File: bindgen-tests/python/tests/interfaces.py
from test_package.uniffi_bindgen_tests import *
import unittest

class TestErrors(unittest.TestCase):
    def test_interfaces(self):
        interface = TestInterface(20)
        self.assertEqual(interface.get_value(), 20)
        self.assertEqual(clone_interface(interface).get_value(), 20)

    def test_records_with_interface_fields(self):
        two = TwoTestInterfaces(first=TestInterface(1), second=TestInterface(2))
        swapped = swap_test_interfaces(two)
        self.assertEqual(swapped.first.get_value(), 2)
        self.assertEqual(swapped.second.get_value(), 1)

    def test_interface_refcounts(self):
        interface = TestInterface(20)
        interface2 = TestInterface(20)
        def func_that_clones_interface(interface):
          return clone_interface(interface)

        interface2_clone = func_that_clones_interface(clone_interface(interface))
        _ = interface.get_value()
        # Check that only the 2 actual references remain after the dust clears
        self.assertEqual(interface.ref_count(), 2)

if __name__ == '__main__':
    unittest.main()
